### ISSUE TITLE
Backend-agnostic invocation pattern for fused MX op from TorchAO

### ISSUE THREAD

**Issue by slabhs-aws** [pytorch/ao#3908]
In TorchAO, `MXLinear` currently does `quantize(x)` + `quantize(w)` + `scaled_mm(...)` (i.e., two quantize calls followed by `scaled_mm`).
Our backend wants this to be fused into one custom kernel: `fused_quantize_mx_scaled_mm`.

For backend-agnostic integration, what is the intended upstream pattern?

1. Add a fused op as an ATen op in PT Core (e.g., `aten::fused_quantize_mx_scaled_mm`) (similar to https://github.com/pytorch/rfcs/issues/86) and let the dispatcher route to backend kernel, or
2. Use an indirection via `torch.accelerator.fused_quantize_mx_scaled_mm` which invokes `torch.library.custom_op::fused_quantize_mx_scaled_mm` mapped to backend kernel

If we pursue (2), is upstreaming this fused entrypoint under `torch.accelerator` aligned with PyTorch design philosophy?

**Comment by jerryzh168** [comment]
cc @vkuzo @danielvegamyhre @slayton58

**Comment by slayton58** [comment]
This doesn't feel like a good use-case -- @danielvegamyhre @vkuzo I'm not as familiar with the AO codebase, but in general `W` should be pre-quantized (either once for inference, or post-optimizer-step for training), and only `x` needs quantization -- is that what AO is doing here (eliding the cast for `W` if already converted)?

**Comment by danielvegamyhre** [comment]
@slayton58 
>W should be pre-quantized (either once for inference, or post-optimizer-step for training

for training we quantize W directly before the GEMM / Grouped GEMM, since (1) it changes with every optimizer step, as you noted, and (2) we need to quantize along a different axis for forward vs backward. Also, for cases where torch native code + torch.compile perf is good enough that we don't need custom kernels, there is the added benefit of inductor being able to fuse the quantization ops with surrounding ops.

@slabhs-aws i think you could also do an inductor pass with pattern matching on the graph, to do subgraph replacement with a node for a custom op (fused impl)? rather than explicitly having an aten op for this in core (seems a bit too niche/custom for a core aten op). I don't quite understand your 2nd idea, could you elaborate, or perhaps someone else can weigh in on that

**Comment by slabhs-aws** [comment]
@danielvegamyhre the second idea is this: https://docs.pytorch.org/tutorials/advanced/custom_ops_landing_page.html

I can define a custom op `fused_quantize_mx_scaled_mm ` in my PrivateUse1 backend, and to invoke it, I can use `torch.<backend>.fused_quantize_mx_scaled_mm`. However, i dont want to upstream that to TorchAO, as that hardcodes to my backend. So one way to abstract that out is to use `torch.accelerator` . If I upstream `fused_quantize_mx_scaled_mm` to `torch.accelerator`, within TorchAO I can invoke `torch.accelerator.fused_quantize_mx_scaled_mm` which will lower to my backend's impl of `fused_quantize_mx_scaled_mm` custom op.

Or, is there a better way to invoke a custom op from "outside" like TorchAO?

**Comment by slayton58** [comment]
> seems a bit too niche/custom for a core aten op

That's understating it a little :) This is not an op that'll make it into core.  

Edit: And for the record, it's not a given that regular mxfp8 quant makes it into core via. ATen or otherwise, or ends up living elsewhere

**Comment by slabhs-aws** [comment]
> it's not a given that regular mxfp8 quant makes it into core via. ATen or otherwise, or ends up living elsewhere

Where do you suggest mxfp8 quant should live if not core ATen, and why?

If the reason for `aten::fused_quantize_mx_scaled_mm` is its too niche, can we generalize it as a fused version of [this](https://github.com/pytorch/rfcs/issues/86)? i.e, something like `fused_quantize_scaled_mm` that can optionally take one or more HP/pre-quantized inputs and conduct the matmul.

**Comment by slayton58** [comment]
mxfp8 quant in particular could live in AO, it could live in MSLK, it could live in a general kernel library, it could live in a quantization-specific kernel library. The issue with quantization right now (and in particular micro-scaled quant like MX and NVFP formats) is that core (ATen) only really wants ops that are (for want of a better word) "settled" in that there is one blessed way of doing it, even if there may be different backend implementations - see things like gemm. 

MX quant is not "settled" - look at the thread, there's debate on rceil vs. RN vs. other ways of calculating the scale factor, alongside both RN and RS roundings for the elements.

> can we generalize it as a fused version of...

You could, but I would very strongly suspect that you a) shouldn't and b) don't actually really want to. As previously mentioned, in real performance cases `W` will be pre-quantized, and the gemm itself doesn't want extra work - mixed-precision gemms as an example suffer from conversion overheads, and that's generally a dequant op, which I would argue is cheaper & easier than quant - no reductions, etc.

**Comment by slabhs-aws** [comment]
@slayton58 @danielvegamyhre  A couple of questions:

1. if `aten::quantize` makes it into PTCore, is a generic `aten::fused_quantize_scaled_mm` not tied to MX still not recommended?

2. TorchAO is tightly coupled with triton to invoke triton's mxfp8_dim0_kernel [here](https://github.com/pytorch/ao/blob/main/torchao/prototype/mx_formats/mx_tensor.py#L569). Are new backends expected to follow a similar pattern and upstream their kernel invocations as such? We especially want to invoke fused quantizeMX+scaledMM kernel in Eager mode. So if we upstream like below, would that be acceptable? 

(Assuming we call ours abc_backend)

```
### MXLinear

input_orig_shape = input_hp.shape
input_hp_r = input_hp.reshape(-1, input_orig_shape[-1])

if is_abc_backend_kernel_supported():
    result = torch.abc_backend.fused_mxfp8_dim0_cast_and_scaled_mm_kernel(...)
else:
    input_mx_r_dim0 = MXTensor.to_mx(
        input_hp_r,
        in_elem_dtype,
        block_size,
        scale_calculation_mode,
        kernel_preference,
        mxfp8_dim0_cast_kernel_choice=mxfp8_dim0_cast_kernel_choice,
    )
    .
    .
    .
```
And similarly in backward.

3. We can make the above backend-agnostic using `torch.accelerator` path. i.e, instead of` torch.abc_backend.fused_mxfp8_dim0_cast_and_scaled_mm_kernel`, we invoke `torch.accelerator.fused_mxfp8_dim0_cast_and_scaled_mm_kernel`, for which we need to add `fused_mxfp8_dim0_cast_and_scaled_mm_kernel` [here](https://github.com/pytorch/pytorch/blob/main/torch/accelerator/__init__.py#L25). What should be the name (`fused_mxfp8_dim0_cast_and_scaled_mm_kernel` vs `fused_quantize_mx_scaled_mm`) can be discussed.

4. Say none of the above options work out. Say our kernel exists in https://github.com/aws-neuron/nki-library. How do we lower from PT Land (TorchAO specifically) to a fused kernel here in Eager mode? Is Inductor pattern matching the only solution? Do Native PyTorch/Eager not offer any other solutions?
